From 68f4265b2d82ba96111fe956b83e6a062c922001 Mon Sep 17 00:00:00 2001
From: Alexander Graf <agraf@csgraf.de>
Date: Wed, 24 Feb 2021 17:49:59 +0100
Subject: [PATCH 20/21] XXX

---
 target/arm/hvf/hvf.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/target/arm/hvf/hvf.c b/target/arm/hvf/hvf.c
index 8762657545..9ad84cf35f 100644
--- a/target/arm/hvf/hvf.c
+++ b/target/arm/hvf/hvf.c
@@ -885,6 +885,7 @@ int hvf_vcpu_exec(CPUState *cpu)
             trace_hvf_unknown_smc(env->xregs[0]);
             env->xregs[0] = -1;
         }
+        advance_pc = true;
         break;
     default:
         cpu_synchronize_state(cpu);
